### PR TITLE
Establish existing render/show event order.

### PR DESCRIPTION
**Pull request by scottaj**

### CODE REVIEW

**Comment by scottaj** [comment]
See discussion on #2355.

File: test/unit/region.spec.js [hunk 1]
```diff
@@ -247,6 +246,22 @@ describe('region', function() {
...
 
+    describe('Render and show event ordering', function() {
+      beforeEach(function() {
+        this.viewEvents = _.map(this.triggerSpy.getCalls(), function(call) {
+          return call.args[0];
+        });
+      });
+
+      it('triggers the before render event before the before show event', function() {
+        expect(_.indexOf(this.viewEvents, 'before:render')).to.be.below(_.indexOf(this.viewEvents, 'before:show'));
```
Comment (jasonLaster) [inline test/unit/region.spec.js]: There's a spy calledBefore assertion that could be coo here

Comment (scottaj) [inline test/unit/region.spec.js]: I'll give it a shot. I tried that when I was first writing these and was having trouble getting it to work properly.

Comment (jasonLaster) [inline test/unit/region.spec.js]: yup. we've got some examples in the codebase
https://github.com/marionettejs/backbone.marionette/blob/master/test/unit/collection-view.spec.js#L171

**Comment by ahumphreys87** [comment]
@scottaj hows it going on this one? Would be great to push this through if you get time

**Comment by jasonLaster** [comment]
Ping.

**Comment by scottaj** [comment]
Sorry, been really busy and lost track of this. I'll try and take a look at it tonight.

**Comment by scottaj** [comment]
@ahumphreys87 @jasonLaster - Updated

File: test/unit/region.spec.js [hunk 2]
```diff
@@ -147,9 +146,11 @@ describe('region', function() {
       this.regionBeforeEmptySpy = this.sinon.spy();
 
       this.view = new this.MyView();
+      this.viewOnBeforeRenderSpy = this.sinon.spy(this.view, 'onBeforeRender');
       this.viewRenderSpy = this.sinon.spy(this.view, 'render');
       this.viewOnBeforeShowSpy = this.sinon.spy(this.view, 'onBeforeShow');
       this.viewOnShowSpy = this.sinon.spy(this.view, 'onShow');
+      this.triggerSpy = this.sinon.spy(this.view, 'trigger');
```
Comment (jasonLaster) [inline test/unit/region.spec.js]: do we need this?

**Comment by jasonLaster** [comment]
one comment, otherwise good

File: test/unit/region.spec.js (hunk 2 above)
Comment (scottaj) [inline test/unit/region.spec.js]: Nope, it's a leftover.

**Comment by scottaj** [comment]
Updated.

**Comment by ahumphreys87** [comment]
:+1:  @marionettejs/marionette-core needs :eyes:

File: test/unit/region.spec.js [hunk 3]
```diff
@@ -122,18 +122,17 @@ describe('region', function() {
         onSwapOut: function() {}
       });
 
-      this.MyView = Backbone.View.extend({
+      this.MyView = Marionette.View.extend({
```
Comment (samccone) [inline test/unit/region.spec.js]: Why not add a new test as compared to changing this one?

This test had the intent to test compatibility with a vanilla Backbone view, and now this test is gone :crying_cat_face:

Comment (scottaj) [inline test/unit/region.spec.js]: I thought it was just a Backbone view for simplicity, not to actually test that case. It seemed to fit in with these existing tests, and I didn't know conventions around making new test blocks.

Comment (samccone) [inline test/unit/region.spec.js]: ah yeah it is a bit confusing in the pits of test land, but this change has a few side effects as to coverage of non Mn Views in regions :fire:

Comment (scottaj) [inline test/unit/region.spec.js]: Ah, is there coverage data somewhere?

I can work on reverting it. I guess I'd like to see where everyone stands on my other outstanding patch as if that is likely to get merged I'll just fix it there.

Comment (samccone) [inline test/unit/region.spec.js]: there is coverage metics, however coverage of external lib support does not exist. It is more of a boolean does it work y/n kind of thing currently

Comment (paulfalgout) [inline test/unit/region.spec.js]: Ah yeah interesting.  `onBeforeRender` would not have worked on a vanilla View.  Seems like the vanilla view test would be much less involved as well.  Would it be easier just to make a simpler test case well documented for testing the vanilla view, or is that just too much extra stuff?

Comment (samccone) [inline test/unit/region.spec.js]: I am fine for any solution that keeps the vanilla tests around

**Pull request merged**